pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -264,6 +261,126 @@ func (r *ReconcileMachineHealthCheck) Reconcile(ctx context.Context, request rec
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileMachineHealthCheck) remediate(ctx context.Context, currentHealthy []target, needRemediationTargets []target, errList []error, m *mapiv1.MachineHealthCheck) []error {
+	// remediate unhealthy
+	for _, t := range needRemediationTargets {
+		klog.V(3).Infof("Reconciling %s: meet unhealthy criteria, triggers remediation", t.string())
+		if m.Spec.RemediationTemplate != nil {
+			errList = r.externalRemediation(ctx, m, t, errList)
+		} else {
+			if err := t.internalRemediation(r); err != nil {
+				klog.Errorf("Reconciling %s: error remediating: %v", t.string(), err)
+				errList = append(errList, err)
+			}
+		}
+
+	}
+
+	for _, t := range currentHealthy {
+		if m.Spec.RemediationTemplate != nil {
+
+			// Get remediation request object
+			obj, err := r.getExternalRemediationRequest(ctx, m, t.Machine.Name)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					continue
+				}
+				klog.Errorf("failed to fetch remediation request for machine %q in namespace %q within cluster %q: %v", t.Machine.Name, t.Machine.Namespace, t.Machine.ClusterName, err)
+			}
+			// Check that obj has no DeletionTimestamp to avoid hot loop
+			if obj.GetDeletionTimestamp() == nil {
```

Comment:
better to use `obj.GetDeletionTimestamp().IsZero()`

Reply:
That seems a bit risky. looks like a major change to the upstream logic.
I believe we risk a nil crush.
Maybe check for both is best ?
obj.GetDeletionTimestamp() == nil || obj.GetDeletionTimestamp().IsZero()

Reply:
no worries, in golang it's not risky :)
no need to check if `obj.GetDeletionTimestamp() == nil`. go knows the type and invokes IsZero() even if `GetDeletionTimestamp()`  returns `nil`, and `IsZero()` handles `nil` correctly.

Having said that, I see now that `GetDeletionTimestamp()` already runs `IsZero()` so I'm good with leaving the code as is